redis/src/cluster_async/mod.rs

```diff
@@ -611,40 +605,61 @@ where
                         }
                         _ => cmd.clone(),
                     };
-                    (addr.to_string(), conn, cmd)
+                    let (sender, receiver) = oneshot::channel();
+                    let addr = addr.to_string();
+                    (
+                        (addr.clone(), receiver),
+                        PendingRequest {
+                            retry: 0,
+                            sender,
+                            info: RequestInfo {
+                                cmd: CmdArg::Cmd {
+                                    cmd,
+                                    routing: CommandRouting::Connection { addr, conn },
+                                    response_policy: None,
+                                },
+                                redirect: None,
+                            },
+                        },
+                    )
                 })
             })
-            .collect();
+            .unzip();
         drop(read_guard);
+        core.pending_requests.lock().unwrap().extend(requests);
 
         let extract_result = |response| match response {
             Response::Single(value) => value,
             Response::Multiple(_) => unreachable!(),
         };
 
-        let run_func = |(_, conn, cmd)| {
-            Box::pin(async move {
-                let conn = conn.await;
-                Ok(extract_result(func(conn, cmd).await?))
-            })
+        let convert_result = |res: Result<RedisResult<Response>, _>| {
+            res.map_err(|_| RedisError::from((ErrorKind::ResponseError, "request wasn't handled")))
+                .and_then(|res| res.map(extract_result))
+        };
+
+        let get_receiver = |(_, receiver): (_, oneshot::Receiver<RedisResult<Response>>)| async {
+            convert_result(receiver.await)
         };
 
         // TODO - once Value::Error will be merged, these will need to be updated to handle this new value.
         let result = match response_policy {
```

Comment:
maybe move it to a separate function - process_results, or something similar? IMO it's too long